### PR TITLE
Tests for EIP-3855: PUSH0 instruction

### PR DESCRIPTION
**Pull request by gumb0**
https://eips.ethereum.org/EIPS/eip-3855

Geth implementation in https://github.com/ethereum/go-ethereum/pull/24039

### CODE REVIEW

File: src/GeneralStateTestsFiller/VMTests/vmTests/push0GasFiller.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  pre:
+    a94f5374fce5edbc8e2a8697c15331677e6ebf0b:
+      balance: 10000000
+      code: ''
+      nonce: '0'
+      storage: {}
+
+    0000000000000000000000000000000000000100:
+      balance: 0
+      code: ':raw 0x5f'
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/VMTests/vmTests/push0GasFiller.yml:21]: save gas before,  do (PUSH0) in lllc  https://github.com/winsvega/solidity (add as a PR here, retesteth docker uses it)
save the gas after
sstore the gas used change into storage.

File: src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        data: ':label 1025_push0'
+        gas:  !!int -1
+        value: !!int -1
+      network:
+        - 'London'
+      result:
+        b94f5374fce5edbc8e2a8697c15331677e6ebf0b:
+          storage:
+            0x01: 0x01
+    # Same expectation, but for London execution fails on 5f opcode, for 3855 it fails with stack overflow
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml:175]: new opcode to be added into opcode tests by Ori. 
https://github.com/ethereum/tests/tree/develop/src/GeneralStateTestsFiller/stBadOpcode
and here https://github.com/ethereum/tests/tree/develop/src/Templates/BadOpcode

ask Ori to add this opcode once the realisation is certain

File: src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        171717171717171717171717171717171717171717171717171717171717171717171717171717171717171717171717171717171717171717171717171717\
+        60019055"
+      nonce: '0'
+      storage: {}
+
+    0000000000000000000000000000000000000300:
+      balance: 0
+      # 1025 push0's
+      code: ":raw 0x\
+        5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f\
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml:76]: what are you trying to break by calling it 1025 times ? 

is there a test cases shit? 
I would check push0 in staticcall to see that it act as normal pushes.
also make sure it does not mess up stack if there is some stack pushed already and then push0 happens.

File: src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    b94f5374fce5edbc8e2a8697c15331677e6ebf0b:
+      balance: 0
+      code: ':yul { sstore(0, call(100000, shr(96, calldataload(0)), 0, 0, 0, 0, 0)) sstore(1, 1) }'
+      nonce: 0
+      storage: {}
+
+    0000000000000000000000000000000000000100:
+      balance: 0
+      # push1(01) push0 sstore
+      code: ':raw 0x60015f55'
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml:28]: would that also obsolite "0x6000"?

Comment (gumb0) [reply src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml:28]: No, PUSH1(0) is still valid

File: src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml (hunk 3 above)
Comment (gumb0) [reply src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml:76]: > what are you trying to break by calling it 1025 times ?

This case comes from https://eips.ethereum.org/EIPS/eip-3855#test-cases
I think it checks that exactly 1 item is pushed on stack, so 1025 of them overflow stack.

Comment (gumb0) [reply src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml:76]: > also make sure it does not mess up stack if there is some stack pushed already and then push0 happens.

how would you check this?

Comment (winsvega) [reply src/GeneralStateTestsFiller/VMTests/vmTests/push0Filler.yml:76]: > > also make sure it does not mess up stack if there is some stack pushed already and then push0 happens.
> 
> how would you check this?

init:
store: { 01 :  10,  00 : 10 }
code:  60015f55

after the exuction the corresponding cell must change to 0

**Comment by gumb0** [comment]
The work continues in another PR https://github.com/ethereum/tests/pull/1033

**Pull request closed without merging**